src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs

```diff
@@ -149,56 +154,46 @@ public override long NextInt64()
+                ulong lowPart = (ulong)randomProduct;
+
+                if (lowPart < (ulong)maxValue)
                 {
-                    // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
-                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
-                    int bits = BitOperations.Log2Ceiling((ulong)maxValue);
-                    while (true)
+                    ulong remainder = ((ulong)0 - (ulong)maxValue) % (ulong)maxValue;
+
+                    while (lowPart < remainder)
                     {
-                        ulong result = NextUInt64() >> (sizeof(ulong) * 8 - bits);
-                        if (result < (ulong)maxValue)
-                        {
-                            return (long)result;
-                        }
+                        randomProduct = (UInt128)(ulong)maxValue * NextUInt64();
+                        lowPart = (ulong)randomProduct;
                     }
                 }
 
-                Debug.Assert(maxValue == 0 || maxValue == 1);
-                return 0;
+                return (long)(randomProduct >> (sizeof(ulong) * 8));
             }
 
             public override long NextInt64(long minValue, long maxValue)
             {
+                Debug.Assert(minValue <= maxValue);
+
                 ulong exclusiveRange = (ulong)(maxValue - minValue);
 
-                if (exclusiveRange <= int.MaxValue)
-                {
-                    return Next((int)exclusiveRange) + minValue;
-                }
+                UInt128 randomProduct = (UInt128)exclusiveRange * NextUInt64();
+                ulong lowPart = (ulong)randomProduct;
 
-                if (exclusiveRange > 1)
+                if (lowPart < exclusiveRange)
                 {
-                    // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
-                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
-                    int bits = BitOperations.Log2Ceiling(exclusiveRange);
-                    while (true)
+                    ulong remainder = ((ulong)0 - exclusiveRange) % exclusiveRange;
+
+                    while (lowPart < remainder)
                     {
-                        ulong result = NextUInt64() >> (sizeof(ulong) * 8 - bits);
-                        if (result < exclusiveRange)
-                        {
-                            return (long)result + minValue;
-                        }
+                        randomProduct = (UInt128)exclusiveRange * NextUInt64();
+                        lowPart = (ulong)randomProduct;
                     }
                 }
 
-                Debug.Assert(minValue == maxValue || minValue + 1 == maxValue);
-                return minValue;
+                return (long)(randomProduct >> (sizeof(ulong) * 8)) + minValue;
```

Comment:
Same here, I think `64` would be clearer than `sizeof(ulong) * 8)`

Reply:
Fixed.